File: leetCode/python/p37_SudokuSolver/lc37.py
#!/usr/bin/python3

import numpy as np

class Solution:
    '''
        Write a program to solve a Sudoku puzzle by filling the empty cells.
    
        A sudoku solution must satisfy all of the following rules:

        Each of the digits 1-9 must occur exactly once in each row.
        Each of the digits 1-9 must occur exactly once in each column.
        Each of the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes of the grid.
        The '.' character indicates empty cells.    
    '''    
    def print_board(self, board):
        
        print(np.matrix(board))

    def find_empty_location(self, board):

        for y in range(9):
            for x in range(9):
                if board[y][x] == '.':
                    return y, x

        return None, None    

    def is_num_possible(self, y, x, n, board):
        
        for i in range(9):
            if board[y][i] == str(n):
                return False

        for i in range(9):
            if board[i][x] == str(n):
                return False

        x0 = (x // 3) * 3
        y0 = (y // 3) * 3

        for i in range(3):
            for j in range(3):
                if board[y0+i][x0+j] == str(n):
                    return False         

        return True

    def solveSudoku(self, board):
        '''
            Do not return anything, modify board in-place instead.
        '''

        y, x = self.find_empty_location(board)
    
        if y is None:
            return True

        for guess in range(1, 10):
            if self.is_num_possible(y, x, guess, board):
                board[y][x] = str(guess)
                if self.solveSudoku(board):
                    return True
             
            board[y][x] = '.'

        return False

if __name__ == '__main__':

    s = Solution()
    
    # Test 
    
    board = [['5','3','.','.','7','.','.','.','.'],
             ['6','.','.','1','9','5','.','.','.'],
             ['.','9','8','.','.','.','.','6','.'],
             ['8','.','.','.','6','.','.','.','3'],
             ['4','.','.','8','.','3','.','.','1'],
             ['7','.','.','.','2','.','.','.','6'],
             ['.','6','.','.','.','.','2','8','.'],
             ['.','.','.','4','1','9','.','.','5'],
             ['.','.','.','.','8','.','.','7','9']]
    
    print("\nStarting board:")
    s.print_board(board)
    
    s.solveSudoku(board)
    
    print("Solved board:")
    s.print_board(board)

File: leetCode/c/p37_SudokuSolver/lc37.c
/* Write a program to solve a Sudoku puzzle by filling the empty cells.
    
A sudoku solution must satisfy all of the following rules:

    - Each of the digits 1-9 must occur exactly once in each row.
    - Each of the digits 1-9 must occur exactly once in each column.
    - Each of the digits 1-9 must occur exactly once in each of the 9 3x3 sub-boxes of the grid.
     
The '.' character indicates empty cells. */

#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include "../../../mylib/mylib.h"

void printSudoku(int board[9][9]) {
    
    for (int i = 0; i < 9; i++)
    {
        for (int j = 0; j < 9; j++)
            printf("%d ",board[i][j]);
        printf("\n");
    }
}

void findEmptyLocation(int board[9][9], int* y, int*x) {
    
    for(int i=0; i<9; i++) {
        for(int j=0; j<9; j++) {
            if(board[i][j] == 0) {
                *y = i;
                *x = j; 
                return;
            }
        }
    }
    
    // If there are no empty location return negative values for x and y
    *y = -1;
    *x = -1;
}

bool isNumberPossible(int board[9][9], int y, int x, int n) {

    int x0 = (x / 3) * 3;
    int y0 = (y / 3) * 3;

    for(int i=0; i<9; i++) {
        if(board[y][i] == n)
            return false;
    }
    
    for(int i=0; i<9; i++) {
        if(board[i][x] == n) 
            return false;
    }
    
    for(int i=0; i<3; i++) {
        for(int j=0; j<3; j++) {
            if(board[y0+i][x0+j] == n)
                return false;
        }
    }

    return true;
}

bool solveSudoku(int board[9][9]) {
    
    int y, x;

    findEmptyLocation(board, &y, &x);

    if(y < 0) 
        return true;        
    
    for(int guess=1; guess < 10; guess++) {
    
        if(isNumberPossible(board, y, x, guess)) {
            board[y][x] = guess;
            if(solveSudoku(board))
                return true;
        }

        board[y][x] = 0;
    }

    return false;
}

int main(void) {
    
    // Test:

    printf("Given sudoku:\n");

    int board[9][9] = {{5, 3, 0, 0, 7, 0, 0, 0, 0},
                       {6, 0, 0, 1, 9, 5, 0, 0, 0},
                       {0, 9, 8, 0, 0, 0, 0, 6, 0},
                       {8, 0, 0, 0, 6, 0, 0, 0, 3},
                       {4, 0, 0, 8, 0, 3, 0, 0, 1},
                       {7, 0, 0, 0, 2, 0, 0, 0, 6},
                       {0, 6, 0, 0, 0, 0, 2, 8, 0},
                       {0, 0, 0, 4, 1, 9, 0, 0, 5},
                       {0, 0, 0, 0, 8, 0, 0, 7, 9}}; 

    printSudoku(board);

    solveSudoku(board);

    printf("\nSolved sudoku:\n");

    printSudoku(board);
        
    return EXIT_SUCCESS;
}
